### ISSUE TITLE
Excessive Captcha e-mails

### ISSUE THREAD

**Issue by panicacid** [davidjameshowell/epcigames-freegames-heroku#10]
Hey there, got everything setup last night, thank you for taking the time to put together such a comprehensive solution and readme!

However.. It worked great for the first few hours but now it's serving no purpose other than spamming me with captcha emails so far, none of which have actually had me solve a captcha. I open it I get the green successfull message, then about 2 or 3 more e-mails less than a minute later. Then silence for about an hour I think? Then we're back at it again.

Can I change the frequency that this runs to reduce these? It seemed to be working great for the first few hours but I fear the frequency at which it's running quickly causes EPIC to flag and require captcha every consecutive login there after.

**Comment by davidjameshowell** [comment]
So assuming you've grabbed within the last 16 days or so, there was a bug that was fixed that solved the issue with too many logins. Luckily this is solved. Keep in mind that this is just a wrapper for the upstream epicgames code as well, but our behaviors with using Heroku can exacerbate it. 

I took another pass since there was a few issues upstream where logins would fail and everything looks good there. The issue you are now seeing however is partly related to the web Dyno. After the first initial pass, it exits as 0 and then the web Dyno sees this as a failure and will attempt to restart the web Dyno again, cause a cyclic issue of the captcha emails.

A note on the Captcha emails - it is a proxy between Epic Games and you, so often times there is actually no real "captcha" to solve, only opening to trigger it. I personally have only actually received one real Captcha before when using it.

tl;dr - I'm going to look to see if I can use a regular Dyno worker for the requests and proxy to that worker to ensure that exit codes 0 and such do not trigger restarts.

**Comment by panicacid** [comment]
Morning Buddy, I set everything up on Friday evening which I presume qualifies for the buggy login issue.

Do I need to do anything to update this? (I presume run the update workflow in github?)

I did notice when watching the output on Heroku it was exiting with 0 but then seeing that as an error, shortly after I made this post but it had done everything it needed to up-until that point so I presumed it was just how it exited.

**Comment by Flotilla42** [comment]
> A note on the Captcha emails - it is a proxy between Epic Games and you, so often times there is actually no real "captcha" to solve, only opening to trigger it. I personally have only actually received one real Captcha before when using it.

After increasing log level, and poking and prodding a bit, it seems like the issue of empty (as in turns green/done once you tap the link, with no actual captcha required) captcha/lots of e-mails in a row comes from not reusing the existing captcha response - every single captcha response I've seen in the log, for the last several hours (and many, many e-mails) ends with the exact same base64 code block. If the heroku instance could just be persuaded to somehow use/pass it along, I'm guessing it would eliminate the blank captchas, and seriously cut down on repeat e-mails.

**Comment by davidjameshowell** [comment]
Hi @Flotilla42  - I believe I actually fixed the root issue which was a bad key being set and retrieved subsequently. It looks like this was a simple bug related to the Redis stdin option being placed after the Redis command, this making it so there was two keys and no a proper get/set method. I have pushed a PR fix which I believe to resolve this. It is not running end to end without an issues after the first initial CAPTCHA email.

Additionally, as a side note. Due to the way that Heroku web workers operate, it will attempt this process three times before entering a crashed state and then eventually "dying" which is what we want. We have to do a roundabout process with the web worker due to the way that Dyno process operates (in attempting to keep alive).

**Comment by Flotilla42** [comment]
@davidjameshowell  
After testing the fix for a few days, I can confirm - blissful silence in my inbox. 
The fix has been running for several days now, and not sent one invalid/blank captcha e-mail, has auto-refreshed the login, and claimed games successfully. I'd consider this one resolved. 

And I'm somewhat aware of how the Heroku worker operates in this case, which usually wouldn't be a problem, except for when this bug occurred, and it resulted in a small avalanche of non-necessary captcha e-mails to get it to go trough the motions. Thankfully no avalanches sent anymore.

**Comment by panicacid** [comment]
Anything I need to do to update / implement the fix?

**Comment by Flotilla42** [comment]
Run the update workflow manually, and solve the first captcha (if it wants you to). Maybe generate a new e-mail token/cookie first. 

Check the log on logDNA or whatever you attached on Heroku, to see if it claims everything ok, and refreshes. Sit back, and enjoy.

**Comment by panicacid** [comment]
Awesome, will do, I presume as long as it doesn't consider a successful exit code a bad one in the logs were golden?

**Comment by Flotilla42** [comment]
Depends on what you define as exit code. I believe in most logic a successful exit code would be 1, but this one returns 0, then starts up again due to the way Heroku is used here. 

Essentially as long as you've got 

> purchasableGames: [] 
> info Exiting... 
> info Warning: Using a password with '-a' or '-u' option on the command line interface may not be safe. 
> info OK 
> info Process exited with status 0 
> info State changed from up to crashed 

As your final log entries, then all should have gone according to plan, and the fact that it does everything at least twice (or more!) is normal. 

> Successfully refreshed login

Should also ideally be toward the top when beginning a new round in the log, although in my experience it's somewhat limited in how valid/useful it is as a predictor of captcha e-mails.

**Comment by chaoscreater** [comment]
I've updated to the latest build. I can see now that the captcha emails are actually solving successfully now. Previously, they would always failed to solve.

However, the cron job schedule is still not applying correctly. I've set mine to run every 168 hours, or every 7 days (i.e every week), instead of running every 4 hours (which is the default I think). I still get more captcha emails than I would like, because the job runs every X hours and every day.

![image](https://user-images.githubusercontent.com/18227319/109441573-7582f100-7a9a-11eb-837a-75e7cf3e2576.png)

![image](https://user-images.githubusercontent.com/18227319/109441575-787de180-7a9a-11eb-91e0-e3c113906ab3.png)

**Comment by panicacid** [comment]
@chaoscreater Aye I changed the cron job for mine but it still runs every 3 to 6 hours by the looks of it.

**Comment by panicacid** [comment]
Hrmm I ran the update workflow on my fork but still getting captcha after captcha every 20 minutes.

Am I missing something?

**Comment by Flotilla42** [comment]
Try updating again, then reset your cookie.

**Comment by chaoscreater** [comment]
Have generated new cookie and reset it in the workflow already.

I'm not getting as many captchas as I used to in the old version. I would used to consistently get one every 2~4 hours or somewhere around that. Now, I'm getting maybe around 4~5 per day. The captcha emails would solve successfully as well, whereas before they would always fail apart from the very first one that you get, right after you reset the cookie workflow.

Still, this is not running as per my configured cron schedule.

**Comment by panicacid** [comment]
Weird, did anybody else wake up to an e-mail from Epic this morning saying their account was locked due to suspicious activity?

**Comment by chaoscreater** [comment]
Didn't get locked, but got an email about suspicious activity.

**Comment by davidjameshowell** [comment]
Can you guys let me know if you are using 2FA authentication or not? I have not received any email from either captcha nor suspicious activity. What does the email specify, if anything beyond "activity"?

**Comment by panicacid** [comment]
@davidjameshowell  so I'm using the 2FA on the account for Epic, went through the setup again to get the mfa setup code etc.

The email says:

Hello ****, 

You’re receiving this email because we noticed suspicious activity on your Epic Games account. Most often, the cause of this type of activity is the result of using the same password on multiple websites, use of a weak password, or phishing.

To keep your account safe, we’ve initiated a password reset, so you’ll need to change the password before logging in again. Below you’ll find helpful tips to ensure your account remains secure! 
  | Use this link to reset your password: |

**Comment by Flotilla42** [comment]
I'm using 2FA as well, and haven't run into any issues whatsoever yet.

**Comment by Flotilla42** [comment]
That lasted for all of a month, now it's back to its old behaviour - attempting to stuff the same captcha response in somewhere, then mail me about a new captcha, generate the same captcha ID again, email me about a new captcha, repeat. Then finally fail when login has been attempted five-six times.

I've tried updating the cookie several times, I've tried solving the captcha on different devices, I've tried updating the cookie secret and then updating, and ..so far it's not done anything. I'm guessing it's a problem with Redis, but I'm too unfamiliar with it to be certain. I did notice that the captcha string length logged changed right after doing the new encoded coookie + update, where it went from one log message + 2 lines in a new message, to almost 2 complete log messages worth of encoded gibberish.

**Comment by egecelikci** [comment]
> That lasted for all of a month, now it's back to its old behaviour

same

**Comment by davidjameshowell** [comment]
Are you guys actually able to set only the email, password, and TOTP code and retry without using the cookie method? I am no longer using the cookie method and have only experienced a handful of captchas in the past weeks.

**Comment by Flotilla42** [comment]
> Are you guys actually able to set only the email, password, and TOTP code and retry without using the cookie method?

I actually haven't tested that, but on a whim I did test something else - I removed the cloned repository, heroku instance, and reinstalled it all (didn't actually take that long), and it's running smoothly again, at least for the last twelve hours. My reasoning was that the same exact Captcha kept repeating, so it was probably not being stored/used properly anymore - wiping all of it was probably overkill, but it does work again. 

Now the interesting question is if it starts breaking itself in one months time, in which case I would suspect an overflow of some sort, that takes that long to trigger.

**Comment by chaoscreater** [comment]
I've reinstalled everything but I can't configure Mailgun. I clicked the addon and the URL is redirected to https://app.mailgun.com/app/dashboard. It will show loading for a bit but eventually stops loading and the page is completely blank. Have tried this on multiple browsers. Can't configure the email domain if I can't see anything. I have done this before and it worked before, not sure what's going on.

**Comment by Flotilla42** [comment]
> n

Try removing mailgun as an addon, then re-add it. Also try adding a logger to your heroku instance, I'd recommend papertrail, as it comes with a historical log, unlike logdna. Could be that you could get the link for the dashboard trough the logs - the error message usually gives it out, although you might have to increase the log level environment variable to see it.

**Comment by chaoscreater** [comment]
Thanks for the suggestion. I have tried it just now and the result is still the same. I get a different URL compared to before (https://addons-sso.heroku.com/apps/90d1802d-85b9-499f-9fad-5b761575876d/addons/2d2a6f56-9c56-40fe-98bc-xxxxxxxxxxxxxxxxxxx), but it still redirects to the Mailgun dashboard.

In both my attempts, I received this popup on the initial page load:

![image](https://user-images.githubusercontent.com/18227319/113782060-b222c600-9785-11eb-927d-94f1f55682c1.png)

I've tried clicking on Learn More and on the close button and I still don't see anything behind the page. Have tried this on a different machine in a different network.

Regarding adding a logger, I'm not too sure how to do that. I've never used Heroku before, but will try to figure it out...

**Comment by Flotilla42** [comment]
I believe it was resources -> find more addons -> logging. Free level is more than good enough.

If all else fails, completely delete the heroku instance, then rebuild/launch it again via GitHub. Should take about ten minutes or so.

**Comment by chaoscreater** [comment]
I have already deleted the Herkoku instance and let Github rebuild it.

The issue isn't related to Heroku. It's to do with Mailgun. Heroku uses SSO to login to Mailgun and it's the Mailgun dashboard display that is the issue...

I've just installed Papertrail and I haven't seen any Mailgun related errors. I have however found this, which doesn't seem good and I'm not sure how to fix it...


```
 Apr 06 16:03:34 epicgamesclaimerchaoscreater app/web.1 [2021-04-06 23:03:34.123 +0000] INFO  (18 on b8deb286-6f37-46e7-b90c-eb6772a72e9f): Checking free games for xxxxxxxxxxxx@gmail.com 
Apr 06 16:03:34 epicgamesclaimerchaoscreater app/web.1 [2021-04-06 23:03:34.150 +0000] ERROR (18 on b8deb286-6f37-46e7-b90c-eb6772a72e9f): Unexpected end of JSON input
Apr 06 16:03:34 epicgamesclaimerchaoscreater app/web.1     SyntaxError: Unexpected end of JSON input
Apr 06 16:03:34 epicgamesclaimerchaoscreater app/web.1         at JSON.parse (<anonymous>)
Apr 06 16:03:34 epicgamesclaimerchaoscreater app/web.1         at Object.newCookieJar (/usr/app/src/common/request.ts:49:29)
Apr 06 16:03:34 epicgamesclaimerchaoscreater app/web.1         at /usr/app/src/index.ts:16:29
Apr 06 16:03:34 epicgamesclaimerchaoscreater app/web.1         at async Promise.all (index 0)
Apr 06 16:03:34 epicgamesclaimerchaoscreater app/web.1         at main (/usr/app/src/index.ts:31:3)
Apr 06 16:03:34 epicgamesclaimerchaoscreater app/web.1 Exiting...
Apr 06 16:03:34 epicgamesclaimerchaoscreater app/web.1 Warning: Using a password with '-a' or '-u' option on the command line interface may not be safe.
Apr 06 16:03:34 epicgamesclaimerchaoscreater app/web.1 OK
Apr 06 16:03:34 epicgamesclaimerchaoscreater heroku/web.1 Process exited with status 0
Apr 06 16:03:34 epicgamesclaimerchaoscreater heroku/web.1 State changed from up to crashed

```

**Comment by davidjameshowell** [comment]
> I have already deleted the Herkoku instance and let Github rebuild it.
> 
> The issue isn't related to Heroku. It's to do with Mailgun. Heroku uses SSO to login to Mailgun and it's the Mailgun dashboard display that is the issue...
> 
> I've just installed Papertrail and I haven't seen any Mailgun related errors. I have however found this, which doesn't seem good and I'm not sure how to fix it...
> 
> ```
>  Apr 06 16:03:34 epicgamesclaimerchaoscreater app/web.1 [2021-04-06 23:03:34.123 +0000] INFO  (18 on b8deb286-6f37-46e7-b90c-eb6772a72e9f): Checking free games for xxxxxxxxxxxx@gmail.com 
> Apr 06 16:03:34 epicgamesclaimerchaoscreater app/web.1 [2021-04-06 23:03:34.150 +0000] ERROR (18 on b8deb286-6f37-46e7-b90c-eb6772a72e9f): Unexpected end of JSON input
> Apr 06 16:03:34 epicgamesclaimerchaoscreater app/web.1     SyntaxError: Unexpected end of JSON input
> Apr 06 16:03:34 epicgamesclaimerchaoscreater app/web.1         at JSON.parse (<anonymous>)
> Apr 06 16:03:34 epicgamesclaimerchaoscreater app/web.1         at Object.newCookieJar (/usr/app/src/common/request.ts:49:29)
> Apr 06 16:03:34 epicgamesclaimerchaoscreater app/web.1         at /usr/app/src/index.ts:16:29
> Apr 06 16:03:34 epicgamesclaimerchaoscreater app/web.1         at async Promise.all (index 0)
> Apr 06 16:03:34 epicgamesclaimerchaoscreater app/web.1         at main (/usr/app/src/index.ts:31:3)
> Apr 06 16:03:34 epicgamesclaimerchaoscreater app/web.1 Exiting...
> Apr 06 16:03:34 epicgamesclaimerchaoscreater app/web.1 Warning: Using a password with '-a' or '-u' option on the command line interface may not be safe.
> Apr 06 16:03:34 epicgamesclaimerchaoscreater app/web.1 OK
> Apr 06 16:03:34 epicgamesclaimerchaoscreater heroku/web.1 Process exited with status 0
> Apr 06 16:03:34 epicgamesclaimerchaoscreater heroku/web.1 State changed from up to crashed
> ```

That end of JSON is most likely related to busted cookies. Can you clear out your cookies var and use email/password/totp?

I'm going to be taking a more deeper dive around that and the Mailgun to see if I can reproduce it.

**Comment by chaoscreater** [comment]
When I saw the latest comments in this thread about how people are able to get it working again without using cookies, I thought I'd try the same. So I had already removed my Github fork and Heroku instance and re-created everything from scratch. Only difference is that this time, I'm not using cookies, I'm just using email & password. I don't even have a TEMPORARY_EMAIL_COOKIE added into Github secret. I'm not using MFA for my Epic Games account either. The only Github Secrets I have added are:

**HEROKU_API_KEY (yoru Heroku API key - can be found in Account Setings -> APi Keys)
APP_NAME (the name of the Heroku application, this must be unqiue across Heroku and will fail if it is not)
EMAIL_ADDRESS (your Epic Games email address)
EPIC_GAMES_PASSWORD (your Epic Games password)**

**Comment by chaoscreater** [comment]
Ignore the Mailgun issue that I mentioned in an earlier post. It is still an issue, but I managed to access the dashboard on my phone. Not sure why it's not working on PC and I've tried it on different PCs and different browsers, not using any addons or anything that could have blocked it. Super weird issue.

The main issue for me is the error log that I posted above. I have no idea how to resolve that and I have redeployed a new Heroku instance again just now...

**Comment by HandyHat** [comment]
I've got the same issue. Providing a cookie leads to infinite captcha emails, and not providing one leads to the JSON error.

**Comment by davidjameshowell** [comment]
I pushed an update which should resolve the JSON error in #15 . However, it appears that many if not most people are still having issues from the upstream project on excessive CAPTCHAs.

Please see the upstream issue here - https://github.com/claabs/epicgames-freegames-node/issues/97

**Comment by davidjameshowell** [comment]
@all I just revisited this and I believe there was a fundamental flaw - you need to make sure that you are gathering the cookie from the `/id` endpoint as per [this instruction](https://github.com/claabs/epicgames-freegames-node#:~:text=While%20viewing%20the%20Epic%20Games%20Store%20page%2C%20open%20the%20EditThisCookie%20extension%20window%2C%20change%20the%20URL%20to%20https%3A//www.epicgames.com/id%2C%20and%20click%20the%20export%20button%3A)

I was able to export my cookie from `/id` and successfully use it to login without needing to solve any captchas.

**Comment by HandyHat** [comment]
> 
> 
> @ALL I just revisited this and I believe there was a fundamental flaw - you need to make sure that you are gathering the cookie from the `/id` endpoint as per [this instruction](https://github.com/claabs/epicgames-freegames-node#:~:text=While%20viewing%20the%20Epic%20Games%20Store%20page%2C%20open%20the%20EditThisCookie%20extension%20window%2C%20change%20the%20URL%20to%20https%3A//www.epicgames.com/id%2C%20and%20click%20the%20export%20button%3A)
> 
> I was able to export my cookie from `/id` and successfully use it to login without needing to solve any captchas.

Doing this did indeed fix it
